### ISSUE TITLE
Convolutional Neural Networks - Part 2

### ISSUE THREAD

**Issue by cavinjallen** [cavinjallen/CSCI373-Capstone#10]
### **The Feature Learning Phase**
In a CNN consists of an arbitrary number of pairs of Convolution and Pooling Layers
Convolution for one Filter in a Convolution Layer
Take a filter of whatever size (5*5*3) and Slide it over the Input array with a Stride of 1
In each step of the slide, take the dot product between each filter element and each element of each subarea of the Input array. For every dot product taken, the result is a scalar. 
The output is then a Feature Map
If the Stride is larger than 1 the Feature Map becomes more reduced. 
**Padding:**
Depending on the size of input array, size of filters and stride, the sliding process can miss to apply the filter to some input array element
A possibility is to ‘pad’ the original input array with a frame and use the extended array as a basis for the convolution (surround with a border of dummy elements (0’s work fine))
If padding is never used, the system of arrays are shrinking rapidly but if padding is used systematically the size of arrays are kept up. 
Repeated convolution for all filters in a convolution layer
Each convolution layer comprises a set of independent Filters
Each filter is independently convolved with the input image
If there are 6 filters in the first convolution layer, this will generate 6 feature maps of shape X*X*1. Stack these maps together to get a “new image” of size X*X*6
**Pooling (subsampling) Layer:**
A Pooling layer is frequently used in a convolutional neural network with the purpose to progressively reduce the spatial size of the representation to reduce the amount of features and the computational complexity of the network. 
Pooling layer operated on each Feature map independently
The main reason for the pooling layer is to prevent the model from overfitting
Common Approaches used in pooling are Max pooling. Ex: MAXPOOL of 2x2 would cause a Filter of 2 by 2 to traverse over the entire matrix with a Stride of 2 and pick the largest element from the window to be included in the next representation map. Average pooling also is used and this one takes the average instead of the max. 
Two aspects of the neuron structures in the convolution and pooling layers:
Weight-sharing: Certain feature/filter should treat all subareas of the visual space similarity the same weights should be employed within a convolution computation phase. Brings down complexity
Local Connectivity (in an ANN any node can be connected to any neuron) but with the layer structure of CNN strictly restrain the connectivity of the neurons within small sub areas
**Flattening:** 
The output of convolution and pooling layers before entering the fully connected layers if flattened.
The dimensions of the input array will be flattened into one large dimension. Ie: a 3D array of 10x10x10 would become a 1D array of 1000 elements when flattened
The Fully Connected Layers:
The FCL takes an input, a flattened array representing the activation maps of high level features from earlier layers and outputs an N dimensional vector
The FCL determines which features best correlate to a particular class.
Activation functions used:
ReLU (Rectified Linear unit) and Leaking ReLU activation functions in the convolution and pooling layers
Others include Sigmoid and Tanh, typically used for FC networks aims for binary classification problems, Can be used for the output layers of a CNN.
Softmax for binary classification but is primarily aimed for the multi class case where the non-normalized output of a network is mapped onto a probability distribution over predicted output classes. Typically used in the output layers of CNN.
